### ISSUE TITLE
[v6 rc3] re-initializing a form that had an initialValues prop supplied to it breaks "pristine"

### ISSUE THREAD

**Issue by fredantell** [erikras/redux-form#1436]
I'm using redux form v6-rc3.

I have a form that I'm passing "initialValues" to. That part works perfectly well.

The form is being used to keep state for a set of search filters. So a typical user will be changing fields, hitting "search," looking at the results, changing the search filters, again and again. 

When the user hits "search" we re-initialize the form with those values and those become the new "pristine." They may also decide to start over by clearing the form completely, and then that would be the new pristine.

To accomplish this, I am importing the initialize action creator from redux-form and then passing it the new values for the form. It fires properly, and the fields update correctly in the UI. However, the pristine prop is still set to false.

When I check the underlying code in redux-form I see this:
https://github.com/erikras/redux-form/blob/v6.0.0-rc.3/src/reduxForm.js#L375-L380

            const { form, getFormState, initialValues } = props
            const stateInitial = getIn(formState, 'initial')
            const initial = initialValues || stateInitial || empty
            const values = getIn(formState, 'values') || initial
            const pristine = deepEqual(initial, values)

So of course my pristine is false because "initial" and "values" are not deeply equal.
However, if initial defaults to always being initialValues, then wouldn't it auto break any reset-ing of the form? You're always going to be comparing against the values you provided as props and never against the up to date items you're attempting to initialize with.

It seems like this could be fixed trivially by giving stateInitial priority in the assignment:
            const initial = stateInitial || initialValues|| empty

but I wouldn't be surprised if I'm just missing something altogether.

**Comment by erikras** [comment]
This is some superb issue reporting. 🏆 I knew immediately what you meant and what the problem was. Your problem is caused by this:
```js
const initial = initialValues || stateInitial || empty
                ^^^^^^^^^^^^^^^^
```
It was trivial to write a test to reproduce your issue.

However...

I can't really juggle the order on those values, because it breaks many other things. Redux Form has to prioritize either the `initialValues` passed in as a prop, or the initial values in the Redux Store, and I think that in *most* use cases, it needs to prioritize the prop.

So what I think this means is that you have to choose. Either update initial values via the prop or via dispatching `INITIALIZE`, but not both.

Make sense?

**Comment by fredantell** [comment]
Thanks for the response erikras. Happy that you had an easy time following my post. Given how hard you've worked on this library and how much I appreciate it, I figured I should at least try to write clearly when asking for help. :)

I thought I would post as a a followup and share my workaround for two reasons. 1) Maybe I'm doing things poorly and someone can correct me. 2) If I'm not, then others can benefit.

My reason for using initialValues was to grab state from the URL. We have a lot of query params in the URL, they're stored in redux state, so I grab them, turn them into an object of initialVales and pass that as a prop. It's really only meant to happen once on load. After that, as the user manipulates the search form, the URL will update to keep in sync with it (that way users can easily share their view of the data with each other by IMing or emailing a link).

I solved my problem essentially by doing what erikras suggested. I constructed the object of initialVaues as normal,but rather than passing it as a prop for reduxForm, I passed it using some other name: "initialFormValues" in my case. Then in the component's constructor, which runs only once, I used the imported initialize action creator to set my form to it. Problem solved.

There was one other caveat in my case. If I had 10 potential fields I read from the URL, many of them can be empty/undefined. ReduxForm was still always getting initialized with an object with 10 keys. The form would render properly. Pristine worked. The Values object was identical and had the same 10 keys, with many of them frequently undefined.

However if I touched the form and edited an undefined field, and then deleted it back out, then rather than resetting it to undefined, the field disappears entirely from the Values object. Now Values and Initial are no longer equal and Pristine is incorrectly set to false again. The solution to that was just to strip out all the undefined keys before initializing the form. They weren't changing what was rendered in the UI anyway, and now "values" and "initial" always stay in sync.
